Zeroing during arming

peabody124 opened the pull request:
This improves the zeroing of the gyros while arming the quadcopter and improves the accelerated convergence during that time.

peabody124 committed:
Attitude: Clean up the initialization code

Use sensibly labeled enum to indicate when the complimentary filter
is initializing or during arming.

peabody124 committed:
Attitude: When zero gyro on arming, use a block averaging of the gyros

Previously an IIR filter was used but this would normally over shoot
if there is much error.  This code is a bit more complicated but
should avoid that problem.

peabody124 committed:
Attitude: Level attitude more smoothly during arming

Previously accelKp was set to 1 for the entire arming period.  This
is overkill and we can use a decayed schedule for a smooth
convergence.

kubark42 commented on flight/Modules/Attitude/coptercontrol/attitude.c:
```diff
@@ -101,11 +115,16 @@
+// For computing the average gyro during arming
+static bool accumulating_gyro = false;
+static uint32_t accumulated_gyro_samples = 0;
+static float accumulated_gyro[3];
```
Putting all the accumulating_gyro_X variables into a structure could make the code cleaner.

```
accumulate_gyro {
  bool accumulating;
  uint32_t num_samples;
  float[3] samples;
}
```

peabody124 committed:
Address some comments from
https://github.com/PhoenixPilot/PhoenixPilot/pull/181

peabody124 committed:
Attitude: Apply the zero gyros while arming to main code

This replicated the CC code.  Adds a clearer set of state labels during
arming and poweron.  Needs two improvements:

1) since the filter can be changed on the fly using the absolute system
time to time initialization is going to create  bugs when switching to
the complimentary filter

2) adds some static variables which I would prefer to eradicate

peabody124 committed:
Increase the default AccelTau to 0.1

peabody124 committed:
Attitude: Move the gyro bias estimation variables into a structure

peabody124 committed:
Attitude: Consolidate the complimentary filter variables into a structure

By moving all these variables into their own structure it will be easier
in the future to make the complimentary filter a standalone library or
allocate the memory required dynamically.

peabody124 committed:
Attitude: Make sure timing is relative to when filter was first run

Since we can toggle between filters on the fly, using TickCount() to
time how long since initialization will not normally be correct.  This
stores the time value when initialized and compares against that.

peabody124 committed:
Simulation: Add a pios_sensors interface

Now that the attitude module checks whether a mag queue is registered
simulation needs to know about pios_sensors to know to use the mag in
the ekf.  This adds this interface.
